### DISCUSSION TITLE
Preserve block data on cancel of PlayerBlockBreakEvent

### DISCUSSION THREAD

**Discussion by zax71** [Minestom/Minestom#1596] [Help]
When just cancelling the event with `event.setCancelled(true);` it works fine for protection, but blocks with data (e.g: signs) will be placed in their default state (as in with no text).

I have tried storing the broken block and replacing it, but that doesn't seam to work.
```java
Block block = event.getBlock();
Point blockPosition = event.getBlockPosition();
instance.setBlock(blockPosition, block);
```

**Comment by emortaldev** [comment]
How have you set the sign data? There's a chance it was only set client side and so the server still just thinks its a regular sign

**Reply by zax71** [reply]
All I did was copy someone else’s block handler for a sign and load an anvil world

**Reply by GreatWyrm** [reply]
I'm fairly certain this a vanilla minecraft bug, signs will lose text when broken, but only client sided.

Try interacting with the block again. From my quick testing (on a vanilla server), trying to break the block again will restore the text.

**Reply by emortaldev** [reply]
> All I did was copy someone else’s block handler for a sign and load an anvil world

there's a chance that was mine lol

**Reply by zax71** [reply]
> > All I did was copy someone else’s block handler for a sign and load an anvil world
> 
> there's a chance that was mine lol

Here it is, I forget where I got it from though xD
```java
public final class Sign implements BlockHandler {

    @Override
    public @NotNull NamespaceID getNamespaceId() {
        return NamespaceID.from("minecraft", "sign");
    }

    public Collection<Tag<?>> getBlockEntityTags() {
        return Arrays.asList(
                Tag.Byte("GlowingText"),
                Tag.String("Color"),
                Tag.String("Text1"),
                Tag.String("Text2"),
                Tag.String("Text3"),
                Tag.String("Text4"));
    }
}
```

**Reply by zax71** [reply]
> I'm fairly certain this a vanilla minecraft bug, signs will lose text when broken, but only client sided.
> 
> Try interacting with the block again. From my quick testing (on a vanilla server), trying to break the block again will restore the text.

Breaking the block twice doesn't seam to change anything.

**Reply by Timofffee** [reply]
> Here it is, I forget where I got it from though xD

I think from here: https://github.com/EmortalMC/Immortal/blob/main/src/main/kotlin/dev/emortal/immortal/blockhandler/SignHandler.kt
